### PR TITLE
Unvendor distutils

### PR DESCRIPTION
**Pull request by rth**
Packages the distutils stdlib module as a standalone package,
```
377K distutils.data
221K distutils.data.br
19K distutils.js
```
to reduce the minimal pyodide size.

Part of #1542 

This also,
 - adds documentation on what's supported or not in the stdlib
 - makes a few changes to `pyodide_build/buildall.py` to be able to add such stdlib packages as a dependency in other packages (which didn't work before). For instance, until there is a release with https://github.com/pypa/packaging/pull/396 micropip has a dependency on distutils.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -156,6 +160,16 @@ build/test.data: $(CPYTHONLIB) $(UGLIFYJS)
...
+build/distutils.data: $(CPYTHONLIB) $(UGLIFYJS)
+	( \
+		cd $(CPYTHONLIB)/distutils; \
+		find . -type d -name __pycache__ -prune -exec rm -rf {} \; \
+	)
+	( \
+		cd build; \
+		python $(FILEPACKAGER) distutils.data --lz4 --preload ../$(CPYTHONLIB)/distutils@/lib/python$(PYMINOR)/distutils --js-output=distutils.js --export-name=pyodide._module --exclude __pycache__ \
+	)
+	$(UGLIFYJS) build/distutils.js -o build/distutils.js
```
Comment (rth) [inline Makefile:172]: Not particularly happy about having to repeat this for each unvendored package, but I'm not sure what else we can do.

I considered creating `meta.yaml` for such stdlib package and treating them as regular package. But that doesn't work as we need to be sure that CPython build artifacts are available (which is not part of the current build / dependency resolution setup)

File: docs/index.rst [hunk 2]
```diff
@@ -25,6 +25,7 @@ Using Pyodide
    usage/serving-pyodide-packages.md
    usage/loading-packages.md
    usage/type-conversions.md
+   usage/wasm-constraints.md
```
Comment (rth) [inline docs/index.rst:28]: I'm open to better names

File: pyodide_build/buildall.py [hunk 3]
```diff
@@ -131,7 +159,11 @@ def generate_dependency_graph(
     while packages:
         pkgname = packages.pop()
 
-        pkg = Package(packages_dir / pkgname)
+        pkg: BasePackage
+        if pkgname in UNVENDORED_STDLIB_MODULES:
+            pkg = StdLibPackage(packages_dir / pkgname)
+        else:
+            pkg = Package(packages_dir / pkgname)
```
Comment (rth) [inline pyodide_build/buildall.py:166]: Most of the code refactoring in this file is to be able to do the above.

File: docs/usage/wasm-constraints.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+Most of the Python standard library is functional, except for the modules
+listed in the sections below. A large part of the CPython test suite passes except for
+tests skipped in
+[`src/tests/python_tests.txt`](https://github.com/pyodide/pyodide/blob/main/src/tests/python_tests.txt)
+or via [patches](https://github.com/pyodide/pyodide/tree/main/cpython/patches).
+
+### Removed modules
+
+Following modules are removed from the standard library to reduce download size and
```
Comment (hoodmane) [inline docs/usage/wasm-constraints.md:14]:
```suggestion
The following modules are removed from the standard library to reduce download size and
```
and similarly in the other sections.

**Comment by rth** [comment]
So distutils ended up being used more frequently than I though in packages. I fixed it for the packages we include though that would still be an issue for micropip installed packages. We may want to have an option in `loadPyodide` to load the full stdlib in the future.

**Comment by hoodmane** [comment]
> So distutils ended up being used more frequently than I though in packages

An unfortunate aspect of standard libraries is that packages use them in a lot of circumstances where they wouldn't add a dependency.

> We may want to have an option in `loadPyodide` to load the full stdlib in the future.

For stuff like this it can be nice to add it as an option with the old way of doing things as the default. I like this approach a lot when it makes sense because it gives an extra opportunity to collect info about the change before committing to it. Changing from opt-in to opt-out and then later removing the option entirely makes two smaller breaking changes out of one bigger one.

**Comment by rth** [comment]
Yes, I agree it's better to keep backward compatibility on stdlib with default options of `loadPyodide`, will look into it.

**Comment by rth** [comment]
OK, it should be now loading the the full stdlib by default. I'll extend this to a few other large and infrequently used modules once this PR is merged.

File: docs/project/changelog.md [hunk 5]
```diff
@@ -12,6 +12,17 @@ substitutions:
 
 ## [Unreleased]
 
+## Standard library
+
+- Following standard library modules are now available as standalone packages
```
Comment (hoodmane) [inline docs/project/changelog.md:17]:
```suggestion
- The following standard library modules are now available as standalone packages
```

File: docs/usage/wasm-constraints.md [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+ - lib2to3
+ - tkinter
+ - turtle.py
+ - turtledemo
+ - venv
+ - webbrowser.py
+
+### Included but not working modules
+
+The following modules can be imported, but are not functional due to the limitations of the WebAssembly VM,
```
Comment (hoodmane) [inline docs/usage/wasm-constraints.md:41]:
```suggestion
The following modules can be imported, but are not functional due to the limitations of the WebAssembly VM:
```

File: src/pyodide.js [hunk 7]
```diff
@@ -17,6 +17,9 @@ globalThis.pyodide = {};
  * Pyodide module as a global object called ``pyodide``.
  * @param {string} config.indexURL - The URL from which Pyodide will load
  * packages
+ * @param {bool} config.fullStdLib - Load the full Python standard library.
+ * Setting this parameter to false would explude following modules: distutils.
```
Comment (hoodmane) [inline src/pyodide.js:21]:
```suggestion
 * Setting this to false excludes following modules: distutils.
```

**Review by hoodmane — APPROVED** [review]
Looks good. Hopefully this process will be simplified as we unvendor more things.

**Pull request merged**